xnmt/encoder.py

```diff
@@ -36,17 +36,27 @@ def transduce(self, embed_sent):
     """
     raise NotImplementedError('Unimplemented transduce for class:', self.__class__.__name__)
 
+  def get_final_state(self):
+    """ Return the state that represents the transduced sequence """
+    return None
+
 class BuilderEncoder(Encoder):
   def transduce(self, embed_sent):
-    out = None
+    states = None
     if hasattr(self.builder, "transduce"):
-      out = self.builder.transduce(embed_sent)
+      states = self.builder.add_inputs(embed_sent)
     elif hasattr(self.builder, "initial_state"):
-      out = self.builder.initial_state().transduce(embed_sent)
+      states = self.builder.initial_state().add_inputs(embed_sent)
     else:
       raise NotImplementedError("Unimplemented transduce logic for class:",
                                 self.builder.__class__.__name__)
-    return ExpressionSequence(expr_list=out, mask=embed_sent.mask)
+    output_expr = [dy.concatenate([state_f.h()[-1], state_b.h()[-1]]) for state_f, state_b in states]
+    self.final_encoder_state = states[-1]
+    return ExpressionSequence(expr_list=output_expr, mask=embed_sent.mask)
+
+  def get_final_state(self):
+    return dy.concatenate([self.final_encoder_state[0].s()[-1], \
```

Comment:
I think in `get_final_state` we should just check the number of elements in final_encoder_state and act appropriately (e.g. just return `self.final_encoder_state[0].s()` in the case of unidirectional nets). Then (as @msperber noted), in `Translator` we can add several options for how to pass the state from encoder to decoder. The simple method we have now will fail in cases where the number of layers or number of elements in the encoder is different than the decoder, but that's fine for now.